### PR TITLE
8294865: x86: Improve the code generation of MulVB and MulVL

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch simplifies and improves the code generation of `MulVB` and `MulVL` nodes,

- MulVB can be implemented by alternating `vmullw` on odd and even-index elements and combining the results.
- MulVL can be implemented on non-avx512dq by computing the product of each 32-bit half and adding the results together.

Vector API benchmark shows the results of `MUL` operations:

                                                  Before                After
    Benchmark          (size)   Mode  Cnt      Score     Error      Score     Error   Units   Change
    Byte64Vector.MUL     1024  thrpt   15   8948.607 ± 194.646   8860.404 ± 203.109  ops/ms   -0.99%
    Byte128Vector.MUL    1024  thrpt   15  12915.839 ± 291.262  13554.662 ± 488.695  ops/ms   +4.95%
    Byte256Vector.MUL    1024  thrpt   15  12129.959 ± 245.710  23279.276 ± 669.725  ops/ms  +91.92%
    Long128Vector.MUL    1024  thrpt   15   1183.663 ±  36.440   1489.892 ±  35.356  ops/ms  +25.87%
    Long256Vector.MUL    1024  thrpt   15   1911.802 ±  95.304   2834.088 ±  77.647  ops/ms  +48.24%

Please have a look and have some reviews, thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8294865](https://bugs.openjdk.org/browse/JDK-8294865): x86: Improve the code generation of MulVB and MulVL

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [51d39f78](https://git.openjdk.org/jdk/pull/10571/files/51d39f78e16a4e1d2343de026c7fdd526e794723)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10571/head:pull/10571` \
`$ git checkout pull/10571`

Update a local copy of the PR: \
`$ git checkout pull/10571` \
`$ git pull https://git.openjdk.org/jdk pull/10571/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10571`

View PR using the GUI difftool: \
`$ git pr show -t 10571`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10571.diff">https://git.openjdk.org/jdk/pull/10571.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
I submitted testing.

**Review by vnkozlov — APPROVED** [review]
My testing passed.

**Comment by vnkozlov** [comment]
You need second review.

File: src/hotspot/cpu/x86/matcher_x86.hpp
Comment (iwanowww) [inline src/hotspot/cpu/x86/matcher_x86.hpp:195]: Why do you unconditionally return `7` here? Is it because AVX512 doesn't feature a vector instruction to multiply byte vectors?

**Review by iwanowww — COMMENTED** [review]
Looks good.

**Review by iwanowww — APPROVED** [review]

File: src/hotspot/cpu/x86/matcher_x86.hpp
Comment (merykitty) [reply src/hotspot/cpu/x86/matcher_x86.hpp:195]: Yes AFAIK there is no instruction to perform byte vector multiplication directly on x86.

**Comment by merykitty** [comment]
Thanks a lot for your reviews and testings. May I integrate now?

**Comment by vnkozlov** [comment]
> Thanks a lot for your reviews and testings. May I integrate now?

Let me test the latest version before you push.

**Comment by vnkozlov** [comment]
@merykitty testing looks good. You can integrate now.

**Pull request closed without merging**
